esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6245,117 +6180,130 @@ void CWsDfuEx::exportRecordDefinitionBinaryType(const char *recordDefinition, Me
         throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in exportBinaryType.");
 }
 
-void CWsDfuEx::getFileAccessBeforePublish(IEspContext &context, const char *fileName, const char *cluster,
-    const char *jobId, MemoryBuffer& layoutBin, IFileDescriptor *fileDesc, IUserDescriptor *udesc,
-    IConstDFUFileAccessRequestBase &req, IEspDFUFileCreateResponse &resp)
-{
-    IEspDFUFileAccessInfo &accessInfo = resp.updateAccessInfo();
-    getFilePartsInfo(context, fileDesc, fileDesc->numParts(), true, accessInfo);
-
-    StringBuffer keyPairName;
-    unsigned port;
-    bool secure;
-    getFileDafilesrvConfiguration(keyPairName, port, secure, cluster);
-
-    StringBuffer metaInfo, expiryTime;
-    getFileMeta(metaInfo, expiryTime, fileName, fileDesc, udesc, jobId, keyPairName, req);
-    accessInfo.setMetaInfoBlob(metaInfo);
-    accessInfo.setExpiryTime(expiryTime);
-
-    if (layoutBin.length() == 0)
-    {
-        if (!req.getReturnJsonTypeInfo() && !req.getReturnBinTypeInfo())
-            return;
-        resp.setWarning("RecordTypeInfoBin or RecordTypeInfoJson requested, but no type info found from ECLRecordDefinition");
-    }
-
-    if (req.getReturnJsonTypeInfo())
-    {
-        StringBuffer jsonLayout;
-        Owned<IRtlFieldTypeDeserializer> deserializer(createRtlFieldTypeDeserializer());
-        const RtlTypeInfo *typeInfo = deserializer->deserialize(layoutBin);
-        dumpTypeInfo(jsonLayout, typeInfo);
-        if (jsonLayout.length())
-            accessInfo.setRecordTypeInfoJson(jsonLayout.str());
-    }
-    if (req.getReturnBinTypeInfo())
-    {
-        MemoryBuffer binLayout;
-        layoutBin.swapWith(binLayout);
-        if (binLayout.length())
-            accessInfo.setRecordTypeInfoBin(binLayout);
-    }
-}
-
 bool CWsDfuEx::onDFUFileCreate(IEspContext &context, IEspDFUFileCreateRequest &req, IEspDFUFileCreateResponse &resp)
 {
     try
     {
-        IConstDFUFileAccessRequestBase &requestBase = req.getRequestBase();
-        SecAccessFlags accessType = translateToSecAccessFlags(requestBase.getAccessType());
-        if (SecAccess_None == accessType)
+        context.ensureFeatureAccess(FEATURE_URL, SecAccess_Write, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
```

Comment:
Can we check if the requested file exists and is locked here?

Reply:
Without locking support from outside of the environment (future work), e.g. the spark client, there won't be any guarantees. i.e. the file might be created after any checking here.
I think the best we can do for now, is complain at publish time (if overwrite=false).